OpenGarage support

swbradshaw opened the pull request:
## Description:
This adds a cover platform for the Open Source WiFi garage opener, [OpenGarage](http://rayshobby.net/introducing-opengarage/).   This is similar to the Garadget platform, but has a few differences:

- Cost.  These are ESP8266 based devices that a DIY person can build and flash for under $10. For the non-tech user, they can be purchased for $50 USD.
- Local. This operates on your local network, and does not connect to a cloud server.
- Distance Sensor. Open/Close status is determined by a distance sensor. In addition to door status, this also allows you to create a sensor in HA that can tell if your vehicle is in the garage or not. (See documentation)

I started with the garadget.py code and created this platform based on that.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2521

## Example entry for `configuration.yaml` (if applicable):
```yaml
    cover:
     platform: opengarage
     covers:
         garage:
           host:  192.168.1.12
           device_key: opendoor
           name:  Left Garage Door

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

swbradshaw committed:
OpenGarage.io support

Cleaned up component and ran lint checking

swbradshaw committed:
Fixing lint errors

swbradshaw committed:
Added supported_features and device_class

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Required(CONF_DEVICEKEY): cv.string,
+    vol.Optional(CONF_NAME): cv.string
```
`vol.Required` should go first then `vol.Optional`, preferably alphabetically ordered. Make it easier to keep the documentation in sync.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+Platform for the opengarage.io cover component.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/cover.opengarage
```
`/` is missing at the end.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup OpenGarage covers."""
+    covers = []
+    devices = config.get(CONF_COVERS, {})
```
No default needed here as `CONF_COVERS` is required and can't be empty.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+    covers = []
+    devices = config.get(CONF_COVERS, {})
+
+    _LOGGER.debug(devices)
```
Please remove that.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+class OpenGarageCover(CoverDevice):
+    """Representation of a OpenGarage cover."""
+
+    # pylint: disable=no-self-use, too-many-instance-attributes
```
`too-many-instance-attributes` is globally disabled already.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+        url = '{}/jc'.format(
+            self.opengarage_url
+            )
+        ret = requests.get(url)
```
Please add a timeout to all requests.

fabaff commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,246 @@
+        """Get latest status."""
+        url = '{}/jc'.format(
+            self.opengarage_url
+            )
```
I guess that `url` fits on one line.

swbradshaw committed:
Added timeout to HTTP Requests and other changes based on feedback.

swbradshaw commented:
@fabaff I've made the changes you requested. Let me know if you see anything else. Thanks!

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,242 @@
+        """Start watcher."""
+        _LOGGER.debug("Starting Watcher for command: %s ", command)
+        if self._unsub_listener_cover is None:
+            self._unsub_listener_cover = track_utc_time_change(
```
So I am kinda confused here.

 - You register a listener that fires every second. Since you don't know where you are in the second, it means that you schedule an update to be fired between 0.001 and 0.999 seconds. That seems kinda arbitrary.
 - On first fire, you call schedule_update_ha_state, the listener gets removed in the update

Home Assistant will already call the update method after it calls a service because `should_poll` is returning True.

So why is this "start_watcher" necessary?

balloob replied:
It seems like you want to keep calling update while the state is in opening/closing. Why not have this logic inside the update method ? When update is done, if certain state, schedule another sync.

Also use `track_point_in_time` so that you don't have to unsub anything.

swbradshaw committed:
Removed watcher. It provided little value and could cause issues if status was stuck in a state.

swbradshaw commented:
@balloob The watcher did what you described - updated the status every second when the status is opening/closing. That said, I went ahead and removed it. It doesn't provide much value, and in the case that the door were to get stuck while opening/closing, the watcher would keep running, which isn't ideal.  Also, its worth mentioning that I didn't write this code - it's verbatim from the Garadget cover.

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+        self._available = True
+
+        # Lets try to get the configured name if not provided.
+        try:
```
This seems a duplication of the update method. Just call `add_devices(covers, True)` in your setup method and we'll call update right there. You can then go ahead in the update method to load the name if it is None.

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+            self.signal = status.get('rssi')
+            self.dist = status.get('dist')
+            self._available = True
+        except (requests.exceptions.ConnectionError,
```
You can just catch `RequestException` which is the parent of all Requests exceptions.

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+                          dict(reason=ex))
+            self._state = STATE_OFFLINE
+        except KeyError as ex:
+            _LOGGER.warning('OpenGarage device %(device)s seems to be offline',
```
A KeyError doesn't mean it's offline, it means that you requested a key on a dictionary that didn't hold that key. However, this can never happen because you use `.get()` which will never raise this error. You can remove this block.

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+            self._state_before_move = self._state
+            self._state = STATE_CLOSING
+            ret = self._push_button()
+            return ret.get('result') == 1
```
there is no need to return anything

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+            self._state_before_move = self._state
+            self._state = STATE_OPENING
+            ret = self._push_button()
+            return ret.get('result') == 1
```
No need to return anything

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+        self._state_before_move = None
+        self._state = STATE_STOPPED
+        ret = self._push_button()
+        return ret.get('result') == 1
```
No need to return anything

balloob commented on homeassistant/components/cover/opengarage.py:
```diff
@@ -0,0 +1,222 @@
+
+    @property
+    def should_poll(self):
+        """No polling needed for a demo cover."""
```
Stale comment. The default for polling is already `True` so you can just remove this whole property.

swbradshaw committed:
Changes based on feedback. Added error checking for invalid device_key.

balloob committed:
Lint

balloob commented:
Thanks! 🐬
